Refactor bindings

liamhuber opened the pull request:
Moving more structure methods over to sub-modules based on where our power is actually coming from.

Outstanding stuff that's making this draft:
- @zendegani: Better handling of the docstrings, i.e. with `wraps` and `wrapped_method.__doc__ = "SOME STRING\n" + wrapped_method.__doc__` 
- @sudarsan-surendralal: Moving the remaining more complex methods over to `AseFactory` where appropriate (e.g. `surface`).
- @liamhuber: making `pyiron_to_ase` (and `pyiron_to_X`) a method of `Atoms`, and similarly look into making `ase_to_pyiron` some sort of `from_ase` method.
- @liamhuber: Failing tests, or at least a first look at these failing tests.

liamhuber committed:
Finish moving factory methods over to sub modules

liamhuber committed:
Protect the bindings

liamhuber committed:
No new name

zendegani committed:
Add full path to import, remove unnecessary import

zendegani committed:
Fix passing docstring for read and ase_read

TypeError: unsupported operand type(s) for +=: 'NoneType' and 'str'

zendegani committed:
Remove staticmethod decorator of ase_bulk and fix its docstring passing

zendegani committed:
Fix passing docstring for aimsgb_info and aimsgb_build

TypeError: unsupported operand type(s) for +=: 'NoneType' and 'str'

zendegani committed:
Add docstring passing -- that were ignored by @wraps -- explicitly to bulk, crystal, read, cut, and stack

zendegani committed:
DRY: Add ase_wrapped_doc method to shorten the implementation of prepending the docstring

zendegani commented:
@liamhuber 
The errors are resolved and it is working now.
Handling/prepending of `docstring` is implemented as 
```python
wrapped_method.__doc__ = func("SOME STRING\n") + wrapped_method.__doc__
```

jan-janssen committed:
Fix unit tests

jan-janssen commented:
```
======================================================================
1996
1997
ERROR: test_cyl_and_spherical_avg (atomistics.volumetric.test_volumetric_data.TestVolumetricData)
1998
1999
----------------------------------------------------------------------
2000
2001
2002
2003
2004
Traceback (most recent call last):
2005
  File "/home/runner/work/pyiron_atomistics/pyiron_atomistics/tests/atomistics/volumetric/test_volumetric_data.py", line 68, in test_cyl_and_spherical_avg
2006
    struct = StructureFactory.ase_bulk("Al")
2007
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pyiron_base/generic/util.py", line 192, in decorated
2008
    return function(*args, **kwargs)
2009
  File "/home/runner/work/pyiron_atomistics/pyiron_atomistics/pyiron_atomistics/atomistics/structure/factory.py", line 82, in ase_bulk
2010
    return self.ase.bulk(*args, **kwargs)
2011
AttributeError: 'str' object has no attribute 'ase'
2012

2013
----------------------------------------------------------------------
```

liamhuber commented:
I'll look at the tests later tonight.

zendegani commented on pyiron_atomistics/atomistics/structure/factories/ase.py:
```diff
@@ -36,19 +36,19 @@ def ase_wrapped_doc(f_name, chain):
     def cut(self, *args, **kwargs):
         s.publication_add(publication_ase())
-        return ase_cut(*args, **kwargs)
+        return ase_to_pyiron(ase_cut(*args, **kwargs))
```
I think as @sudarsan-surendralal and @niklassiemer pointed out in [#76](https://github.com/pyiron/pyiron_atomistics/pull/76#discussion_r579790057), there is no need to use `ase_to_pyiron` for `cut` and `stack`.
```suggestion
        return ase_cut(*args, **kwargs)
```

liamhuber replied:
Yes, that was also my understanding.

zendegani commented on pyiron_atomistics/atomistics/structure/factories/ase.py:
```diff
@@ -36,19 +36,19 @@ def ase_wrapped_doc(f_name, chain):
     def stack(self, *args, **kwargs):
         s.publication_add(publication_ase())
-        return ase_stack(*args, **kwargs)
+        return ase_to_pyiron(ase_stack(*args, **kwargs))
```
```suggestion
        return ase_stack(*args, **kwargs)
```

liamhuber committed:
Update pyiron_atomistics/atomistics/structure/factories/ase.py

Co-authored-by: Ali Zendegani <17050255+zendegani@users.noreply.github.com>

liamhuber committed:
Update pyiron_atomistics/atomistics/structure/factories/ase.py

Co-authored-by: Ali Zendegani <17050255+zendegani@users.noreply.github.com>

liamhuber commented:
> ======================================================================
> 1996
> 1997
> ERROR: test_cyl_and_spherical_avg (atomistics.volumetric.test_volumetric_data.TestVolumetricData)
> 1998
> 1999
> ----------------------------------------------------------------------
> 2000
> 2001
> 2002
> 2003
> 2004
> Traceback (most recent call last):
> 2005
>   File "/home/runner/work/pyiron_atomistics/pyiron_atomistics/tests/atomistics/volumetric/test_volumetric_data.py", line 68, in test_cyl_and_spherical_avg
> 2006
>     struct = StructureFactory.ase_bulk("Al")
> 2007
>   File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pyiron_base/generic/util.py", line 192, in decorated
> 2008
>     return function(*args, **kwargs)
> 2009
>   File "/home/runner/work/pyiron_atomistics/pyiron_atomistics/pyiron_atomistics/atomistics/structure/factory.py", line 82, in ase_bulk
> 2010
>     return self.ase.bulk(*args, **kwargs)
> 2011
> AttributeError: 'str' object has no attribute 'ase'
> 2012
> 
> 2013
> ----------------------------------------------------------------------

Right, ok I understand this now. The problem is that the test is trying to invoke `StructureFactory.ase.bulk("Al")` instead of `StructureFactory().ase.bulk("Al")`.

For now I will patch the unit test, but before we merge this into branch I'd like some feedback -- is that syntax we *want* to support? I'm inclined to say no. If you want to use a factory, instantiate one. Don't just try pulling methods off the class. In general that's a bad idea since many of our factories take instantiation arguments (like a `Project` instance). I'm open to counterpoints though.

liamhuber committed:
Fix volumetric test

liamhuber committed:
Move helper out of class and hide it

I didn't like the pycharm highlighting of the first argument as though it were self, since in our use it's not self.

jan-janssen reviewed (approved):
Looks good to me

liamhuber committed:
One argument and no strings

liamhuber commented:
> Looks good to me

Cool. There's some more touch-ups I want to do. I'll try to finish them tonight and leave it all day tomorrow in a passing state for final feedback prior to merging 😃

liamhuber committed:
Obey our own deprecation

liamhuber committed:
Decorator magic to remove duplication

liamhuber committed:
Streamline header.

Sometimes we don't actually wrap them, e.g. in cut where we're simply relying that the input structures are already our Atoms class and not theirs. Technically correct is the best kind of correct.

liamhuber committed:
Maintain direct creator import

liamhuber committed:
Update structures notebook with new syntax

liamhuber committed:
Merge pull request #99 from pyiron/update_structure_notebook

Update structure notebook
